src/sinks/prometheus/remote_write.rs

```diff
@@ -103,6 +109,57 @@ impl SinkConfig for RemoteWriteConfig {
     }
 }
 
+///
+use crate::sinks::util::batch::{Batch, BatchError, PushResult};
+use std::sync::Arc;
+
+struct EncodeBatch<B, I, O>
+where
+    B: Batch<Output = I>,
+{
+    inner: B,
+    transform: Arc<Box<dyn Fn(I) -> O + Send + Sync>>,
+}
+
+impl<B, I, O> Batch for EncodeBatch<B, I, O>
+where
+    B: Batch<Output = I>,
+{
+    type Input = B::Input;
+    type Output = O;
+
+    fn get_settings_defaults(
+        config: BatchConfig,
+        defaults: BatchSettings<Self>,
+    ) -> Result<BatchSettings<Self>, BatchError> {
+        Ok(B::get_settings_defaults(config, defaults.into())?.into())
+    }
+
+    fn push(&mut self, item: Self::Input) -> PushResult<Self::Input> {
+        self.inner.push(item)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    fn fresh(&self) -> Self {
+        Self {
+            inner: self.inner.fresh(),
+            transform: Arc::clone(&self.transform),
+        }
+    }
+
+    fn finish(self) -> Self::Output {
+        (self.transform)(self.inner.finish())
+    }
+
+    fn num_items(&self) -> usize {
+        self.inner.num_items()
+    }
+}
+///
+
```

Comment:
This looks like it should live in `src/sinks/util/batch` instead of here.

Reply:
Yep, will move to util if an additional batch layer is ok for internal events.